### PR TITLE
CPv3- Add Access Control to Smart Contract

### PR DESCRIPTION
**Pull request by bitfalt**
Solves Issue #42 

- Add`only_owner`, `grant_minter_role`, `revoke_minter_role`, `grant_offsetter_role`, `revoke_offsetter_role` in `contracts/project.cairo` 
- Add the OpenZeppelin RBAC Component to the Project contract
- Add Owner validation to `set_absorptions` and `set_project_carbon` in `components/absorber/carbon_handler.cairo`
- Add Owner validation to `set_public_sale_open`, `set_unit_price`, `set_min_money_amount_per_tx` in `components/minter/mint.cairo`
- Add Minter validation to `mint` and `batch_mint` in `contracts/project.cairo`
- Add Offsetter validation to `offset` and `batch_offset` in `contracts/project.cairo`
- Refactor existing tests to pass with the RBAC Component
- Add tests to ensure proper behaviour of the modified functions with the RBAC Component

### CODE REVIEW

File: tests/test_mint.cairo [hunk 1]
```diff
@@ -246,3 +260,92 @@ fn test_cancel_mint() {
...
+
+    let minter = IMintDispatcher { contract_address: minter_address };
+
+    let amount: u256 = 100;
+
+    minter.set_min_money_amount_per_tx(amount);
+}
+
+#[test]
+fn test_set_min_money_amoount_per_tx_with_owner_role() {
```
Comment (julienbrs) [inline tests/test_mint.cairo]:
```suggestion
fn test_set_min_money_amount_per_tx_with_owner_role() {
```

File: tests/test_mint.cairo [hunk 2]
```diff
@@ -246,3 +260,92 @@ fn test_cancel_mint() {
...
+    let minter = IMintDispatcher { contract_address: minter_address };
+
+    let price: u256 = 100;
+
+    minter.set_unit_price(price);
+}
+
+#[test]
+#[should_panic(expected: 'Caller is not the owner')]
+fn test_set_min_money_amoount_per_tx_without_owner_role() {
```
Comment (julienbrs) [inline tests/test_mint.cairo]:
```suggestion
fn test_set_min_money_amount_per_tx_without_owner_role() {
```

**Review by julienbrs — CHANGES_REQUESTED** [review]
Really nice work, HUGE congrats! :fire:
Some little changes:
- You are doing role assertion like this: ```let isMinter = self.accesscontrol.has_role(MINTER_ROLE, get_caller_address());
            assert(isMinter, 'Only Minter can batch mint');```
Couldn't it be refactored by using `assert_only_role(MINTER)`?
To check owner role, you can also use `assert_only_owner` from the `Ownable` component.
- Can you move the additional tests that you've made to their right place inside each file? For example put `test_set_project_carbon_without_owner_role` with the other `test_set_project_carbon_...` tests.
- For the tests, you always set `owner_address` to the needed roles (owner, offsetter, minter). It makes sense for a test like `test_set_project_carbon` because in production, only the deployer/owner will call `set_project_carbon`. However in production, the `minter` role should be given and used only for the minter contract. So in general, please do the testing by acting like a normal user (without any roles) when it's possible. For example in `test_get_available_money_amount`, after deploying and setting up the project, and after giving `MINTER` role to the minter contract, you should switch to a normal user to do the rest of the testing.

**Comment by bitfalt** [comment]
Thank you for the feedback. I'll start working on the suggested changes ASAP and will mention you once everything is modified.

File: tests/test_mint.cairo (hunk 1 above)
Comment (bitfalt) [reply tests/test_mint.cairo]: Thanks for pointing out the typo! Addressed already.

File: tests/test_mint.cairo (hunk 2 above)
Comment (bitfalt) [reply tests/test_mint.cairo]: Thanks for pointing out the typo! Addressed already.

**Comment by bitfalt** [comment]
@julienbrs all the requested changes have been addressed. 
Please review again and let me know if anything else is necessary or if I skipped something.

**Review by julienbrs — APPROVED** [review]
Amazing work, I'll update a new issue so we can address the "User should be the main caller" logic

**Pull request merged**